airbyte-integrations/connectors/source-amazon-ads/unit_tests/test_report_streams.py

```diff
@@ -0,0 +1,310 @@
+    assert len(metrics) == METRICS_COUNT * len(stream.metrics_map)
+
+
+@responses.activate
+def test_brands_report_stream(test_config):
+    setup_responses(
+        init_response_brands=REPORT_INIT_RESPONSE,
+        status_response=REPORT_STATUS_RESPONSE,
+        metric_response=METRIC_RESPONSE,
+    )
+
+    config = Spec(**test_config)
+    ctx = make_context()
+
+    stream = SponsoredBrandsReportStream(config, ctx, authenticator=mock.MagicMock())
+    stream_slice = {"reportDate": "20210725"}
+    metrics = [m for m in stream.read_records(SyncMode.incremental, stream_slice=stream_slice)]
+    assert len(metrics) == METRICS_COUNT * len(stream.metrics_map)
+
+
+@responses.activate
+def test_display_report_stream_report_generation_failure(test_config):
+    setup_responses(
+        init_response=REPORT_INIT_RESPONSE,
+        status_response=REPORT_STATUS_RESPONSE.replace("SUCCESS", "FAILURE"),
+        metric_response=METRIC_RESPONSE,
+    )
+
+    config = Spec(**test_config)
+    ctx = make_context()
+
+    stream = DisplayReportStream(config, ctx, authenticator=mock.MagicMock())
+    stream_slice = {"reportDate": "20210725"}
+    metrics = [m for m in stream.read_records(SyncMode.incremental, stream_slice=stream_slice)]
+    assert metrics == []
+
+
+@responses.activate
+def test_display_report_stream_init_failure(mocker, test_config):
+    config = Spec(**test_config)
+    ctx = make_context()
+    stream = DisplayReportStream(config, ctx, authenticator=mock.MagicMock())
+    stream_slice = {"reportDate": "20210725"}
+    responses.add(
+        responses.POST, re.compile(r"https://advertising-api.amazon.com/sd/[a-zA-Z]+/report"), json={"error": "some error"}, status=400
+    )
+
+    metrics = [m for m in stream.read_records(SyncMode.incremental, stream_slice=stream_slice)]
+    assert metrics == []
+
+
+@responses.activate
+def test_display_report_stream_init_http_exception(mocker, test_config):
+    mocker.patch("time.sleep", lambda x: None)
+    config = Spec(**test_config)
+    ctx = make_context()
+    stream = DisplayReportStream(config, ctx, authenticator=mock.MagicMock())
+    stream_slice = {"reportDate": "20210725"}
+    responses.add(responses.POST, re.compile(r"https://advertising-api.amazon.com/sd/[a-zA-Z]+/report"), body=ConnectionError())
+
+    with raises(ConnectionError):
+        _ = [m for m in stream.read_records(SyncMode.incremental, stream_slice=stream_slice)]
+    assert len(responses.calls) == 5
+
+
+@responses.activate
+def test_display_report_stream_init_too_many_requests(mocker, test_config):
+    mocker.patch("time.sleep", lambda x: None)
+    config = Spec(**test_config)
+    ctx = make_context()
+    stream = DisplayReportStream(config, ctx, authenticator=mock.MagicMock())
+    stream_slice = {"reportDate": "20210725"}
+    responses.add(responses.POST, re.compile(r"https://advertising-api.amazon.com/sd/[a-zA-Z]+/report"), json={}, status=429)
```

Comment:
Why don't you add this to `setup_responses()` function?

Reply:
It has 429 status code, no point of over complicating setup_responses parameters